elpy-shell.el

```diff
@@ -146,6 +146,25 @@ the code cell beginnings defined here."
 (defvar elpy--shell-last-py-buffer nil
   "Help keep track of python buffer when changing to pyshell.")
 
+(defvar elpy-shell--python-version-cache nil)
+(defvar elpy-shell--python-interpreter-cache nil)
+
+(defun elpy-shell-get-python-version ()
+  "Return the python interpreter version"
+  (if (or
+	   (not (eq elpy-shell--python-interpreter-cache
+				python-shell-interpreter))
+	   (null elpy-shell--python-version-cache))
```

Comment:
These lines check to see if we need to make the expensive call to python --version. It assumes the only way the version can change is if python-shell-interpreter changes. Is that a safe assumption? It works for my workflow. But is there a tool, like pyenv, where the python-shell-interpreter is some wrapper script that calls out to different versions based on env vars or dotfiles or something.

Reply:
If `python-shell-interpreter` is just `"python"`, the python version will depend on the first python binaries found in the PATH (if you use virtualenv for exemple, it can change).
But this can be addressed by storing the full python path instead of just the name : 
``` elisp
 (if (or
	   (not (eq elpy-shell--python-interpreter-cache
				(executable-find python-shell-interpreter)))
	   (null elpy-shell--python-version-cache))
```

The problem with pyvenv is more complicated to solve as even the full path will be the same...
I don't have any solution at the moment.